### PR TITLE
Add control over mbus params

### PR DESCRIPTION
**Pull request by baldersheim**
- num network threads
 - num connections per target
 - tcpnodelay
Defaults are unchanged.

@vekterlior @havardpe  PR

### CODE REVIEW

File: jrt/src/com/yahoo/jrt/Connection.java [hunk 1]
```diff
@@ -89,21 +90,23 @@ private void setState(int state) {
...
         owner.sessionInit(this);
     }
 
-    public Connection(TransportThread parent, Supervisor owner, Spec spec, Object context) {
+    public Connection(TransportThread parent, Supervisor owner, Spec spec, Object context, boolean tcpNoDelay) {
         super(context);
         this.parent = parent;
         this.owner = owner;
         this.spec = spec;
+        this.tcpNoDelay = tcpNoDelay;
```
Comment (vekterli) [inline jrt/src/com/yahoo/jrt/Connection.java]: Do we have any numbers that indicate disabling `TCP_NODELAY` is worth it? We unconditionally enabled nodelay for everything to get rid  of spurious latency spikes incurred by the in-kernel deferred packet sending.

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCNetwork.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ private static int getNumThreads() {
     public RPCNetwork(RPCNetworkParams params, SlobrokConfigSubscriber slobrokConfig) {
         this.slobroksConfig = slobrokConfig;
         identity = params.getIdentity();
-        orb = new Supervisor(new Transport(2));
+        orb = new Supervisor(new Transport(params.getNumNetworkThreads(), params.getOptimization() == RPCNetworkParams.Optimization.LATENCY));
```
Comment (vekterli) [inline messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCNetwork.java]: Consider wrapping in a `shouldEnableNoDelay(params)` method or similar to make semantics more obvious

**Review by vekterli — APPROVED** [review]
:+1: Overall LGTM but I'm slightly wary of letting the Nagle algorithm back into the playground. Deferring final say to @havardpe 🙂

File: jrt/src/com/yahoo/jrt/Connection.java (hunk 1 above)
Comment (baldersheim) [reply jrt/src/com/yahoo/jrt/Connection.java]: I saw significantly lower system cpu when hardcoding it yesterday.
However I did a lot of different tests then, and I am not putting more into it than it is a dimension that should be tested further. And for that to happen I need config control over it.

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCNetwork.java (hunk 2 above)
Comment (baldersheim) [reply messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCNetwork.java]: Fixed

File: container-messagebus/src/main/resources/configdefinitions/container-mbus.def [hunk 3]
```diff
@@ -2,9 +2,24 @@
...
 port int default=0
+
+# Number of connections per target
+numconnectionspertarget int default=1
+
+# Number network threads
+numthreads int default=2
+
+# Optimize for latency, or throughput.
+optimization enum {LATENCY, THROUGHPUT} default=LATENCY
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-messagebus/src/main/resources/configdefinitions/container-mbus.def]: So be it, that was one of my options.

File: container-messagebus/src/main/resources/configdefinitions/container-mbus.def [hunk 4]
```diff
@@ -2,9 +2,24 @@
...
 port int default=0
+
+# Number of connections per target
+numconnectionspertarget int default=1
+
+# Number network threads
+numthreads int default=2
+
+# Optimize for latency, or throughput.
+optimize_for enum {LATENCY, THROUGHPUT} default=LATENCY
```
Comment (havardpe) [inline container-messagebus/src/main/resources/configdefinitions/container-mbus.def]: this feels a bit generic

If it only controls tcp_nodelay, we can just be more specific. If we add more code alternatives choosing between latency and throughput later we might not be able to test their effects separately, or they need more configuration than a single flag.

File: storage/src/vespa/storage/config/stor-communicationmanager.def [hunk 5]
```diff
@@ -33,6 +33,8 @@ mbus.rpctargetcache.ttl double default = 600
 ## Any value below 1 will be 1.
 mbus.num_threads int default=4
 
+mbus.optimize_for enum {LATENCY, THROUGHPUT} default = LATENCY
```
Comment (havardpe) [inline storage/src/vespa/storage/config/stor-communicationmanager.def]: this feels a bit generic

If it only controls tcp_nodelay, we can just be more specific. If we add more code alternatives choosing between latency and throughput later we might not be able to test their effects separately, or they need more configuration than a single flag.

**Review by havardpe — APPROVED** [review]

File: container-messagebus/src/main/resources/configdefinitions/container-mbus.def (hunk 4 above)
Comment (baldersheim) [reply container-messagebus/src/main/resources/configdefinitions/container-mbus.def]: I do not want to expose too much details. We should figure out a set that works for all.

File: storage/src/vespa/storage/config/stor-communicationmanager.def (hunk 5 above)
Comment (baldersheim) [reply storage/src/vespa/storage/config/stor-communicationmanager.def]: I do not want to expose too much details. We should figure out a set that works for all.

**Pull request merged**